src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -333,40 +353,48 @@ private void InnerStartOperationReceiveFrom()
 
         internal unsafe SocketError DoOperationReceiveFrom(SafeCloseSocket handle, out SocketFlags flags)
         {
-            PrepareIOCPOperation();
-
             flags = _socketFlags;
 
-            int bytesTransferred;
-            SocketError socketError;
-            if (_buffer != null)
+            SocketError socketError = SocketError.Success;
+            NativeOverlapped* overlapped = AllocateNativeOverlapped();
+            try
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    ref _wsaBuffer,
-                    1,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                int bytesTransferred;
+
+                if (_buffer != null)
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        ref _wsaBuffer,
+                        1,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+                else
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        _wsaBufferArray,
+                        _bufferListInternal.Count,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+
+                socketError = ProcessIOCPResult(socketError == SocketError.Success, bytesTransferred);
+                return socketError;
             }
-            else
+            finally
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    _wsaBufferArray,
-                    _bufferListInternal.Count,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                FreeNativeOverlappedIfNotPending(overlapped, socketError);
```

Comment:
The pattern we use in SocketPal.Windows.cs to handle the equivalent situation is that we catch, free the overlapped, and rethrow.  See code below.

I don't have a strong opinion about which is better, and we don't need to resolve that as part of this checkin, but it would be nice to pick one and be consistent.

```
            asyncResult.SetUnmanagedStructures(buffer, offset, count, null);
            try
            {
                // This can throw ObjectDisposedException.
                int bytesTransferred;
                SocketError errorCode = Interop.Winsock.WSASend(
                    handle,
                    ref asyncResult._singleBuffer,
                    1, // There is only ever 1 buffer being sent.
                    out bytesTransferred,
                    socketFlags,
                    asyncResult.OverlappedHandle,
                    IntPtr.Zero);

                return asyncResult.ProcessOverlappedResult(errorCode == SocketError.Success, bytesTransferred);
            }
            catch
            {
                asyncResult.ReleaseUnmanagedStructures();
                throw;
            }
```